snapcraft/storeapi/_upload.py

```diff
@@ -92,26 +102,52 @@ def upload_files(binary_filename, config=None):
     updown_url = os.environ.get('UBUNTU_STORE_UPLOAD_ROOT_URL',
                                 UBUNTU_STORE_UPLOAD_ROOT_URL)
     unscanned_upload_url = urljoin(updown_url, 'unscanned-upload/')
-    files = {'binary': open(binary_filename, 'rb')}
 
     result = {'success': False, 'errors': []}
 
     session = get_oauth_session(config)
     if session is None:
-        result['errors'] = ['No valid credentials found.']
+        result['errors'] = [
+            'No valid credentials found. Have you run "snapcraft login"?']
         return result
 
     try:
+        binary_file_size = os.path.getsize(binary_filename)
+        binary_file = open(binary_filename, 'rb')
+        encoder = MultipartEncoder(
+            fields={
+                'binary': ('filename', binary_file, 'application/octet-stream')
+            }
+        )
+
+        # Create a progress bar that looks like: [==  ] 50%
+        progress_bar = ProgressBar(
+            widgets=['Uploading {} '.format(binary_filename),
+                     Bar(marker='=', left='[', right=']'), ' ', Percentage()],
+            maxval=os.path.getsize(binary_filename)).start()
+
+        # This callback simply updates the progress bar as the upload
+        # progresses.
+        def upload_progress_callback(monitor):
+            if monitor.bytes_read <= binary_file_size:
+                progress_bar.update(monitor.bytes_read)
```

Comment:
Could you extract a function for this block? maybe call it get_upload_progress_callback, that receives the binary_filename and returns the callback func.